torch/distributed/optim/zero_redundancy_optimizer.py

```diff
@@ -194,133 +218,134 @@ def _clear_cache(self) -> None:
+            primitives are called depending on the managed parameters and
+            expect all the ranks to participate on the same set of parameters.
         """
-
         super().add_param_group(param_group)
+        # NOTE: The rest of the function assumes that the call to the parent's
+        # `add_param_group()` appends the new parameter group and preserves
+        # the previous parameter-group ordering
+
         if self.initialized:
-            # Force a re-partitioning
+            # Force a re-partitioning of the parameters
             self._clear_cache()
-
             param_groups = self.partition_parameters()[self.rank]
+            # NOTE: All parameters in the old parameter groups should be
+            # assigned to the same ranks so that the local optimizers do not
+            # need to be reinitialized
+
+            # Add the parameters assigned to this rank from the new parameter
+            # group to the local optimizer, if any
             if len(param_groups) == len(self.optim.param_groups) + 1:
                 self.optim.add_param_group(param_groups[-1])
 
             # Update the bucketing strategy accordingly
             if self.parameters_as_bucket_view:
-                self._setup_flat_buffers()
+                self._build_param_buckets()
 
     def consolidate_state_dict(self, to: int = 0) -> None:
         r"""
-        Update the consolidated state_dict list, one per rank.
+        Consolidate a list of ``state_dict`` s (one per rank) on the target
+        rank.
 
         Arguments:
-            to (int): the rank that receives the global states. (default: 0)
+            to (int): the rank that receives the optimizer states (default: 0).
 
-        .. warning: This needs to be called on all replicas
+        .. warning: This needs to be called on all ranks.
         """
-
-        # Sync lr and other attributes in case its been updated
+        # Sync the exposed `param_groups` attributes to the local optimizer in
+        # case they have been updated
         self._sync_param_groups(self.param_groups, self.optim.param_groups)
 
-        empty_messenger = torch.tensor([0], dtype=torch.uint8, device=self._device)
-
-        # Pull the sharded state from all the other replicas
-        # Store all the states in order, rank by rank
-
-        # NOTE: In practice, `broadcast` is used, which is wasteful (gather would have been appropriate)
-        # compatibility issues with some backends make the use of broadcast mandatory for now.
-        # a possible follow up would be to move all sharded state management to RPC RRef
+        # Pull the sharded state from all ranks and store them in rank order
+        empty_messenger = torch.tensor([0], dtype=torch.uint8, device=self._default_device)
 
-        self._all_states = []
+        # NOTE: We wastefully use `broadcast()` (e.g. instead of `gather()`)
+        # due to compatibility issues with some backends; a possible follow-up
```

Comment:
I'm guessing this is referring to NCCL? If so we should mention this here and if we don't already have a gh issue, we should mention the lack of gather support in NCCL and add a code pointer for this to that issue.

Reply:
from the top of my head that was with older Gloo (compatibility matrix was pytorch 1.5/1.6/1.7 & NCCL/Gloo & CPU/GPU  with gloo), maybe that this is worth revisiting ?